### PR TITLE
fix: ipv6 ident

### LINKED ISSUE #10: Support for IPv6 is missing
When parsing IPv6 addresses/networks, the parser thinks it's a DNS name

    Error: failed to parse line, skipping rest ":db8::/64 -j ACCEPT" of the line: "2001" is not a valid DNS name

### PR DESCRIPTION
**Pull request by MEschenbacher**
Resolves #10

### CODE REVIEW

**Review by leonnicolas — COMMENTED** [review]
Thanks a lot. Can you please add a test where an IPv6 address starts with `::`. I think that should be possible in theory.

And also the `(d *DNSOrIP) Set(s string)` in https://github.com/kilo-io/iptables_parser/blob/275b7c93e8a7588940a64c72b9ba462d254ba1bc/parser.go#L209
is still just blindly adding a `/32`. I think we just resolve that with this PR to prevent an IPv6 address from being parsed with a `32` network mask. Maybe we can use https://cs.opensource.google/go/go/+/go1.22.0:src/net/ip.go;l=494 if there is no network mask and then https://cs.opensource.google/go/go/+/go1.22.0:src/net/ip.go;l=52 to find out if it is v4 or v6 and adding the suffix accordingly. wdyt?

Also a test where we parse a complete rule that contains an IPv6 address would increase my confidence a lot. Let me know if I can help you with anything!

**Comment by MEschenbacher** [comment]
Yes sounds all good, will do :)

**Comment by leonnicolas** [comment]
~I think IP.IPv4 is not part of go version 1.16, but we can update it. I the only places are in go.mod and the github action file~

Never mind, I totally misread the function signature of `IP.IPv4`.

I guess we can use `net/netip`'s https://pkg.go.dev/net/netip#Addr.Is4

**Comment by MEschenbacher** [comment]
I've added three more commits to implement/fix issues, however solving the one issue at hand, regarding extending `COLON` to decide whether this is a `COLON` or `IDENT` is not as easy as I thought.

File: parser.go [hunk 1]
```diff
@@ -64,6 +64,12 @@ func (d Policy) String() string {
 	return fmt.Sprintf("%s%s %s", prefix, d.Chain, d.Action)
 }
 
+type Commit struct{}
```
Comment (leonnicolas) [inline parser.go]: Can you make a separate PR for this?

**Review by leonnicolas — COMMENTED** [review]
I think the `COMMIT` and "no counter" commits should be in separate PRs. This way we can merge the IPv6 support quickly and not block it by my slow reviews.

File: parser.go (hunk 1 above)
Comment (MEschenbacher) [reply parser.go]: Done :)

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -1,15 +1,22 @@
...
-    - uses: actions/setup-go@v1
-      with:
-        go-version: '1.16'
-    - run: go test ./...
+      - uses: actions/checkout@main
+      - uses: actions/setup-go@v1
+        with:
+          go-version: '1.16'
+      - run: go test ./...
+  fmt:
```
Comment (leonnicolas) [inline .github/workflows/test.yaml]: thanks

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -1,15 +1,22 @@
...
   test:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@master
-    - uses: actions/setup-go@v1
-      with:
-        go-version: '1.16'
-    - run: go test ./...
+      - uses: actions/checkout@main
+      - uses: actions/setup-go@v1
```
Comment (leonnicolas) [inline .github/workflows/test.yaml]: I think if we already touched this, we can update the actions/setup-go version
```suggestion
      - uses: actions/setup-go@v5
```

File: LICENSE [hunk 4]
```diff
@@ -1,4 +1,4 @@
-Copyright (c) 2021 Leon Löchner
+Copyright (c) 2022 Leon Löchner
```
Comment (leonnicolas) [inline LICENSE]:
```suggestion
Copyright (c) 2024 Leon Löchner
```

File: README.md [hunk 5]
```diff
@@ -1,5 +1,7 @@
 # iptables-parser
 
+[![Documentation](https://godoc.org/github.com/kilo-io/iptables_parser?status.svg)](http://godoc.org/github.com/kilo-io/iptables_parser)
```
Comment (leonnicolas) [inline README.md]: thanks

File: parser.go [hunk 6]
```diff
@@ -371,7 +374,7 @@ func (p *Parser) Parse() (l Line, err error) {
 	case COLON:
 		return p.parseDefault(p.s.scanLine())
 	case EOF:
-		return nil, io.EOF //ErrEOF
+		return nil, io.EOF // ErrEOF
```
Comment (leonnicolas) [inline parser.go]:
```suggestion
		return nil, io.EOF
```

File: parser.go [hunk 7]
```diff
@@ -689,28 +692,12 @@ func (p *Parser) scanIgnoreWhitespace() (tok Token, lit string) {
...
 // previously read, or values larger then BUF_SIZE.
 func (p *Parser) unscan(n int) {
-	if p.buf.n+n >= BUF_SIZE {
+	if p.buf.n+n >= BUFSIZE {
 		panic("size exceeds buffer")
 	}
 	p.buf.n += n
 }
 
-func (p *Parser) unscanIgnoreWhitespace(n int) error {
```
Comment (leonnicolas) [inline parser.go]: right, we never use that. There are some comments that mention this method. I can make a follow up to remove those confusing comments.

File: parser_test.go [hunk 8]
```diff
@@ -799,23 +872,55 @@ func TestParser_Parse(t *testing.T) {
...
 							Values: []string{"200-1000"},
 						},
 					},
 				},
 			},
 			err: nil,
 		},
+		{
+			name: "parse rule with match expression statistic",
+			s:    "-A foo ! -s 192.168.178.2 ! --dst 1.1.1.1 -m statistic --mode random --probability 0.50000000000 --every 10 --packet 5",
```
Comment (leonnicolas) [inline parser_test.go]: hmm this is a bit funny. Now, rules that have a probability of 0.5 are different from rules with 0.50.

I wonder if we can do anything about that?

File: parser_test.go [hunk 9]
```diff
@@ -931,13 +1037,15 @@ func TestParser_Parse(t *testing.T) {
...
 		},
 	} {
-		p := NewParser(strings.NewReader(tc.s))
-		s, err := p.Parse()
-		if !reflect.DeepEqual(tc.r, s) {
-			t.Errorf("%d. %s: %q result mismatch:\n\texp=%v\n\tgot=%v\n\terr=%v", i, tc.name, tc.s, tc.r, s, err)
-		} else if tc.err != err && tc.err.Error() != err.Error() {
-			t.Errorf("%d. %s: %q error mismatch:\n\texp=%v\n\tgot=%v.", i, tc.name, tc.s, tc.err, err)
-		}
+		t.Run(tc.name, func(t *testing.T) {
```
Comment (leonnicolas) [inline parser_test.go]: nice

File: targets.go [hunk 10]
```diff
@@ -29,6 +29,8 @@ func (p *Parser) parseTarget(t *Target) (state, error) {
 	switch lit {
 	case "DNAT":
 		s, err = p.parseDNAT(&t.Flags)
+	case "SNAT":
```
Comment (leonnicolas) [inline targets.go]: i think you need to rebase from our main branch. Your branch is behind

**Review by leonnicolas — CHANGES_REQUESTED** [review]
Thanks very much. I think everything is fine with me. Can you rebase from our main branch first? I see some changes in this PR, that are already in our main branch.

**Review by leonnicolas — APPROVED** [review]

**Pull request merged**